ztools/x509/extensions.go

```diff
@@ -63,9 +63,42 @@ type BasicConstraints struct {
 }
 
 type SubjectAltName struct {
-	DNSNames       []string `json:"dns_names,omitempty"`
-	EmailAddresses []string `json:"email_addresses,omitempty"`
-	IPAddresses    []net.IP `json:"ip_addresses,omitempty"`
+	DirectoryNames []pkix.Name
+	DNSNames       []string
+	EDIPartyNames  []pkix.EDIPartyName
+	EmailAddresses []string
+	IPAddresses    []net.IP
+	OtherNames     []pkix.OtherName
+	RegisteredIDs  []asn1.ObjectIdentifier
+	URIs           []string
+}
+
+type jsonSubjectAltName struct {
+	DirectoryNames []pkix.Name              `json:"directory_names,omitempty"`
+	DNSNames       []string                 `json:"dns_names,omitempty"`
+	EDIPartyNames  []pkix.EDIPartyName      `json:"edi_party_names,omitempty"`
+	EmailAddresses []string                 `json:"email_addresses,omitempty"`
+	IPAddresses    []net.IP                 `json:"ip_addresses,omitempty"`
+	OtherNames     []pkix.OtherName         `json:"other_names,omitempty"`
+	RegisteredIDs  []string                 `json:"registered_ids,omitempty"`
+	URIs           []string                 `json:"uniform_resource_identifiers,omitempty"`
+}
+
+func (san SubjectAltName) MarshalJSON() ([]byte, error) {
+	jsan := jsonSubjectAltName{
```

Comment:
This should have a pointer receiver.

`func (san *SubjectAltName) MarshalJSON()`